src/lib/mac/kmac/kmac.cpp

```diff
@@ -0,0 +1,97 @@
+/*
+* KMAC
+* (C) 2023 Jack Lloyd
+* (C) 2023 Falko Strenzke
+* (C) 2023 René Meusel - Rohde & Schwarz Cybersecurity
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#include <botan/internal/kmac.h>
+
+#include <botan/internal/cshake_xof.h>
+#include <botan/internal/fmt.h>
+#include <botan/internal/keccak_helpers.h>
+
+namespace Botan {
+
+KMAC::~KMAC() = default;
+
+void KMAC::clear() {
+   zap(m_encoded_key);
+   m_cshake->clear();
+}
+
+size_t KMAC::output_length() const {
+   return m_output_bit_length / 8;
+}
+
+Key_Length_Specification KMAC::key_spec() const {
+   // KMAC supports key lengths from zero up to 2²⁰⁴⁰ (2^(2040)) bits:
+   // https://nvlpubs.nist.gov/nistpubs/specialpublications/nist.sp.800-185.pdf#page=28
+   //
+   // However, we restrict the key length to 64 bytes in order to avoid allocation of overly
+   // large memory stretches when client code works with the maximal key length.
+   return Key_Length_Specification(0, 64);
+}
+
+bool KMAC::has_keying_material() const {
+   return !m_encoded_key.empty();
+}
+
+std::string KMAC::provider() const {
+   return m_cshake->provider();
+}
+
+void KMAC::start_msg(const uint8_t nonce[], size_t nonce_len) {
+   assert_key_material_set();
+   m_cshake->start({nonce, nonce_len}, {});
+   m_cshake->update(m_encoded_key);
```

Comment:
I don't think this works given the way the MAC contract is specified. We currently only require `start_msg` if you are providing a nonce. But, the nonce is optional with KMAC. So someone using KMAC without a nonce, and just doing init-update-final, will end up with unkeyed KMAC.

This is missed by the tests because `test_mac.cpp` currently always calls start even if the `iv` is empty. This should be extended so we also (when the iv/nonce is not in use) invoking the MAC without calling `start`. (I'll do a PR)

Reply:
Yikes! 😶

Reply:
#3699

Reply:
I'll rebase and fix after #3699 is merged (potentially tomorrow morning).

Reply:
Rebased and fixed. On that note: `cSHAKE_XOF` has the very same issue: If no-one calls `start()` it won't absorb the prefix (i.e. `N` and `S`) and generate wrong results.

Now, with the XOF API still being fresh, we could simply enforce that any XOF must be `start()`ed, but I think that would be confusing and error prone for things like vanilla SHAKE. I'll fix this in an independent PR.